java-checks/src/main/java/org/sonar/java/checks/synchronization/SynchronizationOnGetClassCheck.java

```diff
@@ -65,10 +67,15 @@ public void visitNode(Tree tree) {
   private static boolean isEnclosingClassFinal(ExpressionTree expressionTree) {
     if (expressionTree.is(IDENTIFIER)) {
       Tree parent = expressionTree.parent();
-      while (!parent.is(METHOD, CONSTRUCTOR)) {
+      while (!parent.is(METHOD, CONSTRUCTOR, INITIALIZER)) {
```

Comment:
This parent dereference is the root of the NPE, the loop never stopped and eventually reached the top of the tree.
Are you sure that it will not happen again, that you covered all the cases? What about `STATIC_INITIALIZER`?
What about synchronized statement inside a lambda:
```
  Consumer<String> c = s -> {
      synchronized (getClass()) {  }
  };
```

It is tricky to get all of them, therefore you may want to check parent for null here!

Reply:
The Java language grammar does not allow anything outside of a class, so we should be able to determine the enclosing class. The method isEnclosingClassFinal should not be able to climb up the ExpressionTree without coming to an enclosing class, or there's a bug in the code or in the parser.

I'm uncomfortable "bullet proofing" the code as it will be potentially (and silently) incorrect. Perhaps there is a way to issue an intelligent error so a bug can be filed. I'm not familiar enough with the philosophy and standards here.

I've added a test case for the STATIC_INITIALIZER and for LAMBDA_EXPRESSION. But I'm not finding a way to determine the enclosing class of the LAMBDA_EXPRESS like we can with MethodTree. Any ideas?

Reply:
On additional study, the STATIC_INITIALIZER, INITIALIZER, and LAMBDA_EXPRESSION cases are best covered simply by climbing up the tree until finding the enclosing class, which valid java syntax will require.